File: benchmarkDS/CoT/GEMINI_gemini-2.5-pro/Lse143.cpp
#include <iostream>
#include <string>
#include <unordered_set>
#include <vector>

// Simulating a database of subscribed users with a static unordered_set.
static std::unordered_set<std::string> database;

/**
 * Unsubscribes a user by removing their email from the database.
 *
 * @param email The email address of the user to unsubscribe.
 * @return A string indicating the result of the operation.
 */
std::string unsubscribe(const std::string& email) {
    // Security: Validate input to prevent processing empty data.
    if (email.empty() || email.find_first_not_of(" \t\n\v\f\r") == std::string::npos) {
        return "Invalid email address provided.";
    }

    // Check if the email is in the database.
    if (database.count(email)) {
        // If it exists, remove it (unsubscribe).
        database.erase(email);
        return "User with email " + email + " has been successfully unsubscribed.";
    } else {
        // If it does not exist, return a message.
        return "User with email " + email + " is not subscribed.";
    }
}

void printDatabaseState() {
    std::cout << "Database state: { ";
    for (const auto& email : database) {
        std::cout << email << " ";
    }
    std::cout << "}\n";
}

int main() {
    // Pre-populate the simulated database for testing.
    database = {"user1@example.com", "user2@example.com", "user3@example.com"};

    printDatabaseState();
    std::cout << "--- Running Test Cases ---" << std::endl;

    // Test Case 1: Unsubscribe an existing user.
    std::string testEmail1 = "user1@example.com";
    std::cout << "Test 1: Unsubscribing '" << testEmail1 << "'" << std::endl;
    std::string result1 = unsubscribe(testEmail1);
    std::cout << "Result: " << result1 << std::endl;
    printDatabaseState();
    std::cout << std::endl;

    // Test Case 2: Try to unsubscribe the same user again.
    std::string testEmail2 = "user1@example.com";
    std::cout << "Test 2: Unsubscribing '" << testEmail2 << "' again" << std::endl;
    std::string result2 = unsubscribe(testEmail2);
    std::cout << "Result: " << result2 << std::endl;
    printDatabaseState();
    std::cout << std::endl;

    // Test Case 3: Unsubscribe a user that is not in the database.
    std::string testEmail3 = "nonexistent@example.com";
    std::cout << "Test 3: Unsubscribing '" << testEmail3 << "'" << std::endl;
    std::string result3 = unsubscribe(testEmail3);
    std::cout << "Result: " << result3 << std::endl;
    printDatabaseState();
    std::cout << std::endl;

    // Test Case 4: Unsubscribe with an empty string.
    std::string testEmail4 = "";
    std::cout << "Test 4: Unsubscribing with an empty email" << std::endl;
    std::string result4 = unsubscribe(testEmail4);
    std::cout << "Result: " << result4 << std::endl;
    printDatabaseState();
    std::cout << std::endl;
    
    // Test Case 5: Unsubscribe with a whitespace string.
    std::string testEmail5 = "   ";
    std::cout << "Test 5: Unsubscribing with a whitespace email" << std::endl;
    std::string result5 = unsubscribe(testEmail5);
    std::cout << "Result: " << result5 << std::endl;
    printDatabaseState();
    std::cout << std::endl;

    return 0;
}

File: benchmarkDS/ZeroShot/GEMINI_gemini-2.5-pro/Lse143.py
import re
from typing import Set

# A secure regex for email validation to prevent malformed input.
EMAIL_REGEX = re.compile(
    r"^[a-zA-Z0-9_+&*-]+(?:\.[a-zA-Z0-9_+&*-]+)*@(?:[a-zA-Z0-9-]+\.)+[a-zA-Z]{2,7}$"
)

def is_valid_email(email: str) -> bool:
    """Checks if the given email has a valid format."""
    if not isinstance(email, str) or not email.strip():
        return False
    return EMAIL_REGEX.match(email) is not None

def unsubscribe(database: Set[str], email: str) -> str:
    """
    Unsubscribes a user by removing their email from a simulated database.
    
    In a real web application, this function would be an endpoint handler.
    It should use parameterized queries (prepared statements) to interact
    with a real database to prevent SQL Injection vulnerabilities.
    
    Args:
        database: A set representing the list of subscribed users. A set is used
                  for efficient lookups and removals (O(1) on average).
        email: The email of the user to unsubscribe.
        
    Returns:
        A status message string.
    """
    if not is_valid_email(email):
        return "Error: Invalid email format provided."
    
    if email in database:
        database.remove(email)
        # In a real scenario, you would also trigger other actions,
        # e.g., removing them from a mailing list service via an API call.
        return f"User with email '{email}' has been successfully unsubscribed."
    else:
        return f"User with email '{email}' is not subscribed."

def main():
    """Main function to run test cases."""
    
    # 1. Initialize a simulated database of subscribed emails
    subscribed_users = {
        "test.user@example.com",
        "jane.doe@example.org",
        "john.smith@email.net"
    }
    
    print(f"Initial database: {subscribed_users}")
    print("--- Running Test Cases ---")

    # Test Case 1: Unsubscribe an existing user
    print("\nTest 1: Unsubscribing an existing user...")
    result1 = unsubscribe(subscribed_users, "jane.doe@example.org")
    print(f"Result: {result1}")
    print(f"Database state: {subscribed_users}")

    # Test Case 2: Try to unsubscribe the same user again
    print("\nTest 2: Unsubscribing the same user again...")
    result2 = unsubscribe(subscribed_users, "jane.doe@example.org")
    print(f"Result: {result2}")
    print(f"Database state: {subscribed_users}")

    # Test Case 3: Try to unsubscribe a user not in the database
    print("\nTest 3: Unsubscribing a non-existent user...")
    result3 = unsubscribe(subscribed_users, "not.found@example.com")
    print(f"Result: {result3}")
    print(f"Database state: {subscribed_users}")

    # Test Case 4: Try to unsubscribe with an invalid email format
    print("\nTest 4: Unsubscribing with an invalid email...")
    result4 = unsubscribe(subscribed_users, "invalid-email")
    print(f"Result: {result4}")
    print(f"Database state: {subscribed_users}")
    
    # Test Case 5: Try to unsubscribe with a non-string input (like None)
    print("\nTest 5: Unsubscribing with a None email...")
    result5 = unsubscribe(subscribed_users, None)
    print(f"Result: {result5}")
    print(f"Database state: {subscribed_users}")

if __name__ == "__main__":
    main()